src/foundry/client/pixi/core/shapes/source-polygon.d.ts

```diff
@@ -1,110 +1,247 @@
+
+    /**
+     * Compute the aggregate bounding box which is the intersection of all boundary shapes.
+     * Round and pad the resulting rectangle by 1 pixel to ensure it always contains the origin.
+     */
+    protected _defineBoundingBox(): PIXI.Rectangle;
+
+    /**
+     * Apply a constraining boundary shape to an existing PointSourcePolygon.
+     * Return a new instance of the polygon with the constraint applied.
+     * The new instance is only a "shallow clone", as it shares references to component properties with the original.
+     * @param constraint          - The constraining boundary shape
+     * @param intersectionOptions - Options passed to the shape intersection method
+     * @returns A new constrained polygon
+     */
+    applyConstraint(
+      constraint: PIXI.Circle | PIXI.Rectangle | PIXI.Polygon,
+      intersectionOptions?: Record<string, unknown>,
+    ): PointSourcePolygon;
+
+    /** {@inheritDoc} */
+    contains(x: number, y: number): boolean;
+
+    /**
+     * Constrain polygon points by applying boundary shapes.
+     */
+    protected _constrainBoundaryShapes(): void;
+
+    /**
+     * Test whether a Ray between the origin and destination points would collide with a boundary of this Polygon.
+     * A valid wall restriction type is compulsory and must be passed into the config options.
+     * @param origin      - An origin point
+     * @param destination - A destination point
+     * @param config      - The configuration that defines a certain Polygon type
+     * @param mode        - The collision mode to test: "any", "all", or "closest"
+     *                      (default: "all")
+     * @returns The collision result depends on the mode of the test:
+     *          * any: returns a boolean for whether any collision occurred
+     *          * all: returns a sorted array of PolygonVertex instances
+     *          * closest: returns a PolygonVertex instance or null
+     */
+    static testCollision<Mode extends PointSourcePolygon.CollisionModes>(
+      origin: Point,
+      destination: Point,
+      {
+        mode,
+        ...config
+      }: {
+        /**
+         * The collision mode to test: "any", "all", or "closest"
+         * (default: "all")
+         */
+        mode?: Mode;
+
+        /** The configuration that defines a certain Polygon type */
+        config?: PointSourcePolygonConfig;
+      },
+    ): Mode extends "any" ? boolean : Mode extends "closest" ? PolygonVertex : PolygonVertex[] | null;
+
+    /**
+     * Determine the set of collisions which occurs for a Ray.
+     * @param ray  - The Ray to test
+     * @param mode - The collision mode being tested
+     * @returns The collision test result
+     */
+    protected abstract _testCollision<Mode extends PointSourcePolygon.CollisionModes>(
+      ray: Ray,
+      mode: Mode,
+    ): Mode extends "any" ? boolean : Mode extends "closest" ? PolygonVertex : PolygonVertex[] | null;
```

Comment:
It'd be a little bit cleaner if you do a helper type. The name is up in the air but perhaps `PointSourcePolygon.TestCollision<Mode>`

Here's how I'd implement it:
```ts
type CollisionTypes = {
    any: boolean;
    closest: PolygonVertex;
    all: PolygonVertex[] | null;
} satisfies Record<PointSourcePolygon.CollisionModes, unknown>;

type TestCollision<Mode extends PointSourcePolygon.CollisionModes> = CollisionTypes[Mode];
```

This is of course more verbose but in this case I think it's worth it.